src/coreclr/src/vm/callcounting.cpp

```diff
@@ -0,0 +1,1735 @@
+                    {
+                        break;
+                    }
+                }
+
+                if (!activeCodeVersion.GetILCodeVersion().HasAnyOptimizedNativeCodeVersion(activeCodeVersion))
+                {
+                    GetAppDomain()
+                        ->GetTieredCompilationManager()
+                        ->AsyncPromoteToTier1(activeCodeVersion, scheduleTieringBackgroundWorkRef);
+                }
+                methodDesc->SetCodeEntryPoint(codeEntryPoint);
+                callCountingInfo->SetStage(CallCountingInfo::Stage::Complete);
+                return true;
+            } while (false);
+
+            callCountingStub = callCountingInfo->GetCallCountingStub();
+            if (callCountingStub != nullptr)
+            {
+                break;
+            }
+        }
+        else
+        {
+            _ASSERTE(activeCodeVersion.GetOptimizationTier() == NativeCodeVersion::OptimizationTier0);
+
+            // If the tiering delay is active, postpone further work
+            if (GetAppDomain()
+                    ->GetTieredCompilationManager()
+                    ->TrySetCodeEntryPointAndRecordMethodForCallCounting(methodDesc, codeEntryPoint))
+            {
+                return true;
+            }
+
+            CallCount callCountThreshold = (CallCount)g_pConfig->TieredCompilation_CallCountThreshold();
+            _ASSERTE(callCountThreshold != 0);
+            _ASSERTE(callCountThreshold <= MaximumCallCountThreshold);
+
+            NewHolder<CallCountingInfo> callCountingInfoHolder = new CallCountingInfo(activeCodeVersion, callCountThreshold);
+            callCountingManager->m_callCountingInfoByCodeVersionHash.Add(callCountingInfoHolder);
+            callCountingInfo = callCountingInfoHolder.Extract();
+        }
+
+        if (!g_pConfig->TieredCompilation_UseCallCountingStubs())
+        {
+            // Call counting is not yet complete, so reset or don't set the code entry point to continue counting calls
+
+            if (wasMethodCalled)
+            {
+                return false;
+            }
+
+            // This path is reached after activating a code version when publishing its code entry point. The method may
+            // currently be pointing to the code entry point of a different code version, so an explicit reset is necessary.
+            methodDesc->ResetCodeEntryPoint();
+            return true;
+        }
+
+        callCountingStub =
+            callCountingManager->m_callCountingStubAllocator.AllocateStub(
+                callCountingInfo->GetRemainingCallCountCell(),
+                codeEntryPoint);
+        callCountingInfo->SetCallCountingStub(callCountingStub);
+    } while (false);
+
+    PCODE callCountingCodeEntryPoint = callCountingStub->GetEntryPoint();
+    if (methodDesc->MayHaveEntryPointSlotsToBackpatch())
+    {
+        // The call counting stub should not be the entry point that is called first in the process of a call
```

Comment:
Forwarder stubs are only created for methods that don't already have a precode and currently that's only virtual and interface methods.

Actually after call counting is complete, the forwarder stub can be removed from the hash table since the method typically won't be counted anymore (only if a profiler creates a new IL code version, then it may create another forwarder stub). Also realized that the hash table doesn't need to contain the `MethodDesc` since the forwarder stub already does, so I'll go ahead and make those changes.

Ignoring hash table storage for the moment, I agree it's not too profitable for virtual/interface methods on x64 (saves only 8 bytes). It's more profitable for non-virtual/interface methods on x64 and x86, and for all methods on arm64 and arm32.

Stubs, `CallCountingInfos`, and hash table storage will be there for methods that haven't yet reached the threshold, and until the 4 K completed count is reached (trigger for deleting). I'll post numbers on memory separately.

Regarding your second question, the temporary entry points that are reused for precodes for regular methods (these have a more compact storage scheme), for virtual/interface methods they need to always point to the prestub in order to discover new vtable slots over the same MethodDesc. I thought about using FuncPtrStubs (they are just a precode as well, not any better optimized for storage), and I didn't find a good way of mixing their current use with my intended use. It might be something that can be optimized later with more careful thought.

Reply:
> I thought about using FuncPtrStubs and I didn't find a good way of mixing their current use with my intended use

Could you add some comments in the code describing what the mismatch is? It wasn't immediately obvious to me why it wasn't a viable scenario and someone else seeing the same code later may wonder the same thing.

Reply:
> they are just a precode as well, not any better optimized for storage

The FuncPtr dictionary only stores a Precode*, not a MethodDesc+Precode* pair as this code was doing. I think you noticed the same optimization so once you do that then storage efficiency would match.

Reply:
Ah ok, makes sense

Reply:
Will add a comment. It many not be that beneficial to merge them anyway, as for virtual/interface methods `FuncPtrStubs` are only created for devirtualized cases or cases where a single entry point is needed for multi-call purposes to a virtual/interface method.